<a href="https://colab.research.google.com/github/lurahayes/programming/blob/main/Data_Cleaning_FP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Cleaning for DS3001 Proj 1

## Importing Data From Web

In order to get the necessary data off of the GSS website, we used their [subsetting tool]( https://sda.berkeley.edu/sdaweb/analysis/?dataset=gss21rel3). This allowed us to get a subsetted dataframe of the variables we wanted to examine for each year. Then, we exported these files into our Github, and cleaned them individually. During this process, we took care of missing values by converting them into null values. Finally, we merged the two sets together for analysis.

## Importing Neccessary Packages and Git connection

First, we imported neccessary packages and the link to the data from Github. Then, we imported the data file and converted it to a csv. We removed the commas from the dataset to ensure they would format properly throughout analysis.

In [1]:
!git clone https://github.com/jacqueline-chao/project-1.git

Cloning into 'project-1'...
remote: Enumerating objects: 65, done.
remote: Counting objects: 100% (65/65), done.
remote: Compressing objects: 100% (49/49), done.
remote: Total 65 (delta 28), reused 39 (delta 9), pack-reused 0
Receiving objects: 100% (65/65), 1.53 MiB | 6.71 MiB/s, done.
Resolving deltas: 100% (28/28), done.


In [2]:
import numpy as np

In [3]:
import pandas as pd

In [4]:
file_path_2018 = './project-1/data/2018_data.txt'

In [5]:
file_path_2021 = './project-1/data/2021_data.txt'

In [6]:
df_2018 = pd.read_csv(file_path_2018, delimiter=',')

In [7]:
df_2021 = pd.read_csv(file_path_2021, delimiter=',')

## Cleaning 2018 Data

First, we examined the first few lines and the shape.

In [8]:
df_2018.head()

conmedic partyid  region  id age  sex  race  degree
0       .i       5       1   1  43    1     1       2
1        1       2       1   2  74    2     1       1
2        1       4       1   3  42    1     1       3
3        3       2       1   4  63    2     1       3
4        1       6       1   5  71    1     2       4

In [9]:
df_2018.shape

(2348, 8)

It should be noted that there are more entries for 2021 than 2018.

We found the codes for missing values listed in the GSS codebook.

The following code line examines the number of missing values for the id variable.

In [10]:
miss_val_codes = ['.d', '.i', '.j', '.m', '.n', '.p', '.q', '.r', '.s', '.u', '.x', '.y', '.z']

var = 'id'
df_2018[df_2018[var].isin(miss_val_codes)]

Empty DataFrame
Columns: [conmedic, partyid, region, id, age, sex, race, degree]
Index: []

None of the 2018 respondents had missing ids.

The following code line examines the number of missing values for the conmedic variable.

In [11]:
var = 'conmedic'
df_2018[var].unique().tolist()

['.i', ' 1', ' 3', ' 2', '.n', '.d']

The unidentified values are .i, .n, and .d. The codebook defines them as the following:


*   .i: IAP (Inapplicable) - respondent was asked NOT to respond
*   .n: No answer
* .d: Don't know

We decided to turn these missing values into nans in order to continue our investigation.



In [12]:
miss_val_codes = ['.i', '.n', '.d']
for code in miss_val_codes:
    df_2018[var] = df_2018[var].replace(code, np.nan)

Checking success of value transformation

In [13]:
df_2018[var].unique().tolist()

[nan, ' 1', ' 3', ' 2']

In [14]:
df_2018[var].value_counts()

 2    790
 1    553
 3    211
Name: conmedic, dtype: int64

Now we must examine the total number of missing values for conmedic.

In [15]:
len(df_2018) - np.sum(df_2018[var].value_counts().values)

794

There are 794 missing values for conmdeic.

Next, we cleaned the partyid variable.

In [16]:
var = 'partyid'
df_2018[var].unique().tolist()

[' 5', ' 2', ' 4', ' 6', ' 0', ' 3', ' 1', '.n', ' 7']

Since .n means no answer was given, we must convert it into a nan.

In [17]:
df_2018[var] = df_2018[var].replace('.n', np.nan)

Checking success of transformation

In [18]:
df_2018[var].unique().tolist()

[' 5', ' 2', ' 4', ' 6', ' 0', ' 3', ' 1', nan, ' 7']

In [19]:
df_2018[var].value_counts()

 3    414
 0    379
 1    352
 2    307
 5    272
 4    259
 6    255
 7     77
Name: partyid, dtype: int64

Total number of missing values for partyid identification

In [20]:
len(df_2018) - np.sum(df_2018[var].value_counts().values)

33

There are 33 missing values for partyid.

Now, we must clean the region variable.

In [21]:
var = 'region'
df_2018[var].unique().tolist()

[1, 2, 9, 3, 5, 7, 8, 4, 6]

In [22]:
df_2018[var].value_counts()

5    513
3    388
9    343
7    277
2    232
8    183
6    163
4    125
1    124
Name: region, dtype: int64

There are no unidentified values for region.

Then we looked at the age variable, and converted .n (no answer) values to nans.  

In [23]:
var = 'age'
df_2018[var].unique().tolist()

['43',
 '74',
 '42',
 '63',
 '71',
 '67',
 '59',
 '62',
 '55',
 '34',
 '61',
 '44',
 '41',
 '75',
 '30',
 '40',
 '29',
 '37',
 '56',
 '82',
 '68',
 '20',
 '89',
 '60',
 '65',
 '45',
 '50',
 '52',
 '46',
 '53',
 '22',
 '33',
 '23',
 '28',
 '27',
 '64',
 '79',
 '32',
 '35',
 '21',
 '47',
 '70',
 '77',
 '69',
 '48',
 '81',
 '78',
 '54',
 '58',
 '76',
 '39',
 '38',
 '25',
 '49',
 '18',
 '19',
 '26',
 '57',
 '51',
 '36',
 '72',
 '24',
 '88',
 '66',
 '84',
 '80',
 '31',
 '83',
 '73',
 '86',
 '.n',
 '85',
 '87']

In [24]:
df_2018[var] = df_2018[var].replace('.n', np.nan)

In [25]:
df_2018[var].unique().tolist()

['43',
 '74',
 '42',
 '63',
 '71',
 '67',
 '59',
 '62',
 '55',
 '34',
 '61',
 '44',
 '41',
 '75',
 '30',
 '40',
 '29',
 '37',
 '56',
 '82',
 '68',
 '20',
 '89',
 '60',
 '65',
 '45',
 '50',
 '52',
 '46',
 '53',
 '22',
 '33',
 '23',
 '28',
 '27',
 '64',
 '79',
 '32',
 '35',
 '21',
 '47',
 '70',
 '77',
 '69',
 '48',
 '81',
 '78',
 '54',
 '58',
 '76',
 '39',
 '38',
 '25',
 '49',
 '18',
 '19',
 '26',
 '57',
 '51',
 '36',
 '72',
 '24',
 '88',
 '66',
 '84',
 '80',
 '31',
 '83',
 '73',
 '86',
 nan,
 '85',
 '87']

In [26]:
df_2018[var].value_counts()

34    65
39    55
56    50
30    50
53    48
      ..
79     9
81     9
88     8
83     7
87     5
Name: age, Length: 72, dtype: int64

In [27]:
len(df_2018) - np.sum(df_2018[var].value_counts().values)

7

Finally, we found that are 7 missing values for age.

The sex variable did not contain unidentified variables.

In [28]:
var = 'sex'
df_2018[var].unique().tolist()

[1, 2]

In [29]:
df_2018[var].value_counts()

2    1296
1    1052
Name: sex, dtype: int64

The race variable did not contain unidentified values.

In [30]:
var = 'race'
df_2018[var].unique().tolist()

[1, 2, 3]

In [31]:
df_2018[var].value_counts()

1    1693
2     385
3     270
Name: race, dtype: int64

For degree, we had to convert missing values into nans and find the total number of nans throughout the variable.

In [32]:
var = 'degree'
df_2018[var].unique().tolist()

[2, 1, 3, 4, 0]

In [33]:
miss_val_codes = ['.d', '.s']
for code in miss_val_codes:
    df_2018[var] = df_2018[var].replace(code, np.nan)

In [34]:
df_2018[var].unique().tolist()

[2, 1, 3, 4, 0]

In [35]:
len(df_2018) - np.sum(df_2018[var].value_counts().values)

0

There were 0 missing values in the degree variable.

#### Cleaning 2021 Data

We repeated this process for the 2021 dataset.

In [36]:
df_2021.head()

conmedic partyid  region  id age sex race degree
0        1       3       1   1  65   2    1      1
1        1       2       1   2  60   1    1      3
2       .i       0       2   3  .n  .n    1      0
3        2       3       2   4  .n  .n    1      0
4       .i       0       1   6  .i   2    2      1

In [37]:
df_2021.shape

(4032, 8)

To begin cleaning the 2021 dataset, we ensured there were no missing ids.

In [38]:
miss_val_codes = ['.d', '.i', '.j', '.m', '.n', '.p', '.q', '.r', '.s', '.u', '.x', '.y', '.z']

var = 'id'
df_2021[df_2021[var].isin(miss_val_codes)]

Empty DataFrame
Columns: [conmedic, partyid, region, id, age, sex, race, degree]
Index: []

Then, we examined the conmedic variable for 2021. It included .i, .s, and .d (along with the valid responses). Their meaning is as follows:

*   .i: IAP (Inapplicable) - respondent was asked NOT to respond
*   .s: Skipped on web
* .d: Don't know

We converted them to nan and counted the number of missing values.

In [39]:
var = 'conmedic'
df_2021[var].unique().tolist()

[' 1', '.i', ' 2', ' 3', '.s', '.d']

In [40]:
miss_val_codes = ['.i', '.s', '.d']
for code in miss_val_codes:
    df_2021[var] = df_2021[var].replace(code, np.nan)

In [41]:
df_2021[var].unique().tolist()

[' 1', nan, ' 2', ' 3']

In [42]:
df_2021[var].value_counts()

 2    1346
 1    1070
 3     246
Name: conmedic, dtype: int64

In [43]:
len(df_2021) - np.sum(df_2021[var].value_counts().values)

1370

There are 1370 missing values for conmedic.

The partyid variable for 2021 contained a .n (no answer), so we converted it to nan and found the total number of missing values.

In [44]:
var = 'partyid'
df_2021[var].unique().tolist()

[' 3', ' 2', ' 0', ' 1', ' 5', ' 7', ' 4', ' 6', '.n']

In [45]:
df_2021[var] = df_2021[var].replace('.n', np.nan)

In [46]:
df_2021[var].unique().tolist()

[' 3', ' 2', ' 0', ' 1', ' 5', ' 7', ' 4', ' 6', nan]

In [47]:
df_2021[var].value_counts()

 0    822
 3    817
 1    541
 6    524
 2    471
 5    384
 4    327
 7    114
Name: partyid, dtype: int64

In [48]:
len(df_2021) - np.sum(df_2021[var].value_counts().values)

32

There are 32 missing values for partyid.

The region variable did not contain missing values.

In [49]:
var = 'region'
df_2021[var].unique().tolist()

[1, 2, 9, 3, 5, 7, 8, 4, 6]

In [50]:
df_2021[var].value_counts()

5    800
3    676
9    584
7    426
2    414
8    345
4    314
6    270
1    203
Name: region, dtype: int64

The age variable contained .n (no answer) and .i (inapplicable), so we converted them to nan.

In [51]:
var = 'age'
df_2021[var].unique().tolist()

['65',
 '60',
 '.n',
 '.i',
 '33',
 '20',
 '76',
 '61',
 '37',
 '23',
 '71',
 '21',
 '75',
 '56',
 '22',
 '31',
 '55',
 '53',
 '43',
 '44',
 '32',
 '30',
 '52',
 '34',
 '62',
 '42',
 '29',
 '51',
 '24',
 '26',
 '28',
 '57',
 '45',
 '64',
 '82',
 '58',
 '89',
 '78',
 '50',
 '72',
 '73',
 '49',
 '63',
 '41',
 '66',
 '74',
 '70',
 '46',
 '79',
 '38',
 '68',
 '54',
 '59',
 '25',
 '40',
 '67',
 '84',
 '36',
 '87',
 '35',
 '69',
 '85',
 '39',
 '47',
 '27',
 '48',
 '80',
 '18',
 '81',
 '77',
 '83',
 '19',
 '86',
 '88']

In [52]:
miss_val_codes = ['.n', '.i']
for code in miss_val_codes:
    df_2021[var] = df_2021[var].replace(code, np.nan)

In [53]:
df_2021[var].unique().tolist()

['65',
 '60',
 nan,
 '33',
 '20',
 '76',
 '61',
 '37',
 '23',
 '71',
 '21',
 '75',
 '56',
 '22',
 '31',
 '55',
 '53',
 '43',
 '44',
 '32',
 '30',
 '52',
 '34',
 '62',
 '42',
 '29',
 '51',
 '24',
 '26',
 '28',
 '57',
 '45',
 '64',
 '82',
 '58',
 '89',
 '78',
 '50',
 '72',
 '73',
 '49',
 '63',
 '41',
 '66',
 '74',
 '70',
 '46',
 '79',
 '38',
 '68',
 '54',
 '59',
 '25',
 '40',
 '67',
 '84',
 '36',
 '87',
 '35',
 '69',
 '85',
 '39',
 '47',
 '27',
 '48',
 '80',
 '18',
 '81',
 '77',
 '83',
 '19',
 '86',
 '88']

In [54]:
df_2021[var].value_counts()

67    89
33    82
59    81
57    79
62    77
      ..
19    14
86    14
87     5
88     5
18     4
Name: age, Length: 72, dtype: int64

In [55]:
len(df_2021) - np.sum(df_2021[var].value_counts().values)

333

There are 333 missing values for age.

Sex contained unidentified values .n (no answer), .i (inapplicable), and .s (skipped on web). We converted them to nan and found the total of missing values.

In [56]:
var = 'sex'
df_2021[var].unique().tolist()

[' 2', ' 1', '.n', '.i', '.s']

In [57]:
miss_val_codes = ['.n', '.i', '.s']
for code in miss_val_codes:
    df_2021[var] = df_2021[var].replace(code, np.nan)

In [58]:
df_2021[var].unique().tolist()

[' 2', ' 1', nan]

In [59]:
df_2021[var].value_counts()

 2    2204
 1    1736
Name: sex, dtype: int64

In [60]:
len(df_2021) - np.sum(df_2021[var].value_counts().values)

92

There are 92 missing values for sex

Race contained .i (inapplicable), so we converted it to nan and counted the total number of missing values.

In [61]:
var = 'race'
df_2021[var].unique().tolist()

[' 1', ' 2', ' 3', '.i']

In [62]:
df_2021[var] = df_2021[var].replace('.i', np.nan)

In [63]:
df_2021[var].unique().tolist()

[' 1', ' 2', ' 3', nan]

In [64]:
len(df_2021) - np.sum(df_2021[var].value_counts().values)

54

There are 54 missing values for race.

Degree contained .d (don't know) and .s (skipped on web). We converted these into nans and found the total missing values.

In [65]:
var = 'degree'
df_2021[var].unique().tolist()

[' 1', ' 3', ' 0', ' 4', ' 2', '.d', '.s']

In [66]:
miss_val_codes = ['.d', '.s']
for code in miss_val_codes:
    df_2021[var] = df_2021[var].replace(code, np.nan)

In [67]:
df_2021[var].unique().tolist()

[' 1', ' 3', ' 0', ' 4', ' 2', nan]

In [68]:
df_2021[var].value_counts()

 1    1597
 3    1036
 4     760
 2     370
 0     246
Name: degree, dtype: int64

In [69]:
len(df_2021) - np.sum(df_2021[var].value_counts().values)

23

There were 23 missing values for degree.

## Merging Cleaned 2018 & 2021 Sets

In order to merge the 2018 data set and the 2021 dataset, we added a year column to both. Then, we combined them based on their common ids.

In [70]:
common_ids = pd.merge(df_2018, df_2021, on='id', how='inner')['id'].values.tolist()
len(common_ids)

2136

In [71]:
df_2018_filtered = df_2018[df_2018['id'].isin(common_ids)].copy()
df_2021_filtered = df_2021[df_2021['id'].isin(common_ids)].copy()

In [72]:
df_2018_filtered.head()

conmedic partyid  region  id age  sex  race  degree
0      NaN       5       1   1  43    1     1       2
1        1       2       1   2  74    2     1       1
2        1       4       1   3  42    1     1       3
3        3       2       1   4  63    2     1       3
5      NaN       2       1   6  67    2     1       3

In [73]:
df_2018_filtered['year'] = 2018

In [74]:
df_2018_filtered.head()

conmedic partyid  region  id age  sex  race  degree  year
0      NaN       5       1   1  43    1     1       2  2018
1        1       2       1   2  74    2     1       1  2018
2        1       4       1   3  42    1     1       3  2018
3        3       2       1   4  63    2     1       3  2018
5      NaN       2       1   6  67    2     1       3  2018

In [75]:
df_2021_filtered.head()

conmedic partyid  region  id  age  sex race degree
0        1       3       1   1   65    2    1      1
1        1       2       1   2   60    1    1      3
2      NaN       0       2   3  NaN  NaN    1      0
3        2       3       2   4  NaN  NaN    1      0
4      NaN       0       1   6  NaN    2    2      1

In [76]:
df_2021_filtered['year'] = 2021

In [77]:
df_2021_filtered.head()

conmedic partyid  region  id  age  sex race degree  year
0        1       3       1   1   65    2    1      1  2021
1        1       2       1   2   60    1    1      3  2021
2      NaN       0       2   3  NaN  NaN    1      0  2021
3        2       3       2   4  NaN  NaN    1      0  2021
4      NaN       0       1   6  NaN    2    2      1  2021

In [78]:
complete_df = pd.concat([df_2018_filtered, df_2021_filtered], axis=0)

In [79]:
complete_df.head()

conmedic partyid  region  id age sex race degree  year
0      NaN       5       1   1  43   1    1      2  2018
1        1       2       1   2  74   2    1      1  2018
2        1       4       1   3  42   1    1      3  2018
3        3       2       1   4  63   2    1      3  2018
5      NaN       2       1   6  67   2    1      3  2018

In [80]:
complete_df.shape

(4272, 9)

Finally, we converted the finished dataset into a CSV for analysis and visualization.

In [81]:
complete_df.to_csv('./project-1/data/2018_v_2021_data.csv', index=False)